zip-proportional-output-fee-mechanism-pofm.rst

```diff
@@ -0,0 +1,262 @@
+* marginal_fee = 1000 adapted from @madars' proposal.
+* marginal_fee = 2500 in @daira's proposal.
+* marginal_fee = 1000 for Shielded, Shielding and De-shielding
+  transactions, and marginal_fee = 10000 for Transparent transactions
+  adapted from @nighthawk24's proposal.
+
+(In @madars' and @nighthawk24's original proposals, there was an additional
+`base_fee` parameter that caused the relationship between fee and number
+of inputs/outputs to be non-proportional above the `grace_window_size`. This
+is no longer expressible with the formula specified above.)
+
+Rationale for logical actions
+'''''''''''''''''''''''''''''
+
+A previous proposal used `inputs + outputs` instead of logical actions.
+This would have disadvantages Orchard transactions, as a result of an
+Orchard Action combining an input and an output. The effect of this
+combining is that Orchard requires padding of either inputs or outputs
+to ensure that the number of inputs and outputs are the same. Usage of
+Sapling and transparent protocols does not require this padding, and
+so this could have effectively discriminated against Orchard.
+
+
+Security and Privacy considerations
+===================================
+
+Non-standard transaction fees may reveal specific users or wallets or wallet
+versions, which would reduce privacy for those specific users and the rest
+of the network. However, the advantage of faster deployment argued against
+synchronizing the change in wallet behaviour at a specific block height.
+
+Long term, the issue of fees needs to be re-visited in separate future
+proposals as the blocks start getting consistently full. New ZIPs with
+scaling solutions, will need to be evaluated and applied.
+
+Denial of Service Vulnerability
+-------------------------------
+
+A transaction-rate-based denial of service attack occurs when an attacker
+generates enough transactions over a window of time to prevent legitimate
+transactions from being mined, or to hinder syncing blocks for full nodes
+or miners.
+
+There are two primary protections to this kind of attack in Zcash: the
+block size limit, and transaction fees. The block size limit ensures that
+full nodes and miners can keep up with the blockchain even if blocks are
+completely full. However, users sending legitimate transactions may not
+have their transactions confirmed in a timely manner.
+
+Variable fees could mitigate this kind of denial of service: if there are
+more transactions available than can fit into a single block, then a miner
+will typically choose the transactions that pay the highest fees. If
+legitimate wallets were to increase their fees during this condition, the
+attacker would also increase the fees of their transactions. It is
+sometimes argued that this would impose a cost to the attacker that would
+limit the time window for which they can continue the attack. However, there
+is little evidence that the actual costs involved would be a sufficient
+disincentive.
```

Comment:
Consider deleting some of this section, since the premise of this ZIP is that there is at least some chance that it will work to discourage transaction DoS.